{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.11.11","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":11972829,"sourceType":"datasetVersion","datasetId":7529091},{"sourceId":12293149,"sourceType":"datasetVersion","datasetId":7747972},{"sourceId":12311127,"sourceType":"datasetVersion","datasetId":7759814}],"dockerImageVersionId":31040,"isInternetEnabled":false,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"!pip install biopython","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T16:57:56.403658Z","iopub.execute_input":"2025-06-28T16:57:56.403950Z","iopub.status.idle":"2025-06-28T17:01:19.164142Z","shell.execute_reply.started":"2025-06-28T16:57:56.403922Z","shell.execute_reply":"2025-06-28T17:01:19.163163Z"}},"outputs":[{"name":"stdout","text":"\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f0c4fe4aed0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f0c4fe5a690>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f0c4fe5d550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f0c5008db90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f0c4fe3f990>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[31mERROR: Could not find a version that satisfies the requirement biopython (from versions: none)\u001b[0m\u001b[31m\n\u001b[0m\u001b[31mERROR: No matching distribution found for biopython\u001b[0m\u001b[31m\n\u001b[0m","output_type":"stream"}],"execution_count":1},{"cell_type":"markdown","source":"# **1.1 Loading datasets and data filtration**","metadata":{}},{"cell_type":"code","source":"import pandas as pd\nimport os \n\n# Listing files in the dataset\nos.listdir('/kaggle/input/cancergenelist/')\n\n# Loading the data from a loaded dataset\noncokb_df = pd.read_csv(\"/kaggle/input/cancergenelist/cancerGeneList.tsv\", sep=\"\\t\")\n\nprint(\"Number of genes before filtration: \", len(oncokb_df))\n\n# Show first rows to chceck if the format is correct\noncokb_df.head()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T17:15:34.802232Z","iopub.execute_input":"2025-06-28T17:15:34.802814Z","iopub.status.idle":"2025-06-28T17:15:35.184687Z","shell.execute_reply.started":"2025-06-28T17:15:34.802783Z","shell.execute_reply":"2025-06-28T17:15:35.183858Z"}},"outputs":[{"name":"stdout","text":"The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.\nNumber of genes before filtration:  1192\n","output_type":"stream"},{"execution_count":2,"output_type":"execute_result","data":{"text/plain":"  Hugo Symbol  Entrez Gene ID   GRCh37 Isoform   GRCh37 RefSeq  \\\n0        ABL1              25  ENST00000318560     NM_005157.4   \n1        AKT1             207  ENST00000349310  NM_001014431.1   \n2         ALK             238  ENST00000389048     NM_004304.4   \n3       AMER1          139285  ENST00000330258     NM_152424.3   \n4         APC             324  ENST00000257430     NM_000038.5   \n\n    GRCh38 Isoform   GRCh38 RefSeq Is Oncogene Is Tumor Suppressor Gene  \\\n0  ENST00000318560     NM_005157.4         Yes                       No   \n1  ENST00000349310  NM_001014431.1         Yes                       No   \n2  ENST00000389048     NM_004304.4         Yes                       No   \n3  ENST00000374869     NM_152424.3          No                      Yes   \n4  ENST00000257430     NM_000038.5          No                      Yes   \n\n   # of occurrence within resources (Column J-P) OncoKB Annotated MSK-IMPACT  \\\n0                                              7              Yes        Yes   \n1                                              7              Yes        Yes   \n2                                              7              Yes        Yes   \n3                                              7              Yes        Yes   \n4                                              7              Yes        Yes   \n\n  MSK-HEME FOUNDATION ONE FOUNDATION ONE HEME Vogelstein COSMIC CGC (v99)  \\\n0      Yes            Yes                 Yes        Yes              Yes   \n1      Yes            Yes                 Yes        Yes              Yes   \n2      Yes            Yes                 Yes        Yes              Yes   \n3      Yes            Yes                 Yes        Yes              Yes   \n4      Yes            Yes                 Yes        Yes              Yes   \n\n                            Gene Aliases  \n0                       ABL, JTK7, c-ABL  \n1        AKT, PKB, PRKBA, RAC, RAC-alpha  \n2                                  CD246  \n3  FAM123B, FLJ39827, RP11-403E24.2, WTX  \n4                         DP2.5, PPP1R46  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Hugo Symbol</th>\n      <th>Entrez Gene ID</th>\n      <th>GRCh37 Isoform</th>\n      <th>GRCh37 RefSeq</th>\n      <th>GRCh38 Isoform</th>\n      <th>GRCh38 RefSeq</th>\n      <th>Is Oncogene</th>\n      <th>Is Tumor Suppressor Gene</th>\n      <th># of occurrence within resources (Column J-P)</th>\n      <th>OncoKB Annotated</th>\n      <th>MSK-IMPACT</th>\n      <th>MSK-HEME</th>\n      <th>FOUNDATION ONE</th>\n      <th>FOUNDATION ONE HEME</th>\n      <th>Vogelstein</th>\n      <th>COSMIC CGC (v99)</th>\n      <th>Gene Aliases</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ABL1</td>\n      <td>25</td>\n      <td>ENST00000318560</td>\n      <td>NM_005157.4</td>\n      <td>ENST00000318560</td>\n      <td>NM_005157.4</td>\n      <td>Yes</td>\n      <td>No</td>\n      <td>7</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>ABL, JTK7, c-ABL</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>AKT1</td>\n      <td>207</td>\n      <td>ENST00000349310</td>\n      <td>NM_001014431.1</td>\n      <td>ENST00000349310</td>\n      <td>NM_001014431.1</td>\n      <td>Yes</td>\n      <td>No</td>\n      <td>7</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>AKT, PKB, PRKBA, RAC, RAC-alpha</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ALK</td>\n      <td>238</td>\n      <td>ENST00000389048</td>\n      <td>NM_004304.4</td>\n      <td>ENST00000389048</td>\n      <td>NM_004304.4</td>\n      <td>Yes</td>\n      <td>No</td>\n      <td>7</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>CD246</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>AMER1</td>\n      <td>139285</td>\n      <td>ENST00000330258</td>\n      <td>NM_152424.3</td>\n      <td>ENST00000374869</td>\n      <td>NM_152424.3</td>\n      <td>No</td>\n      <td>Yes</td>\n      <td>7</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>FAM123B, FLJ39827, RP11-403E24.2, WTX</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>APC</td>\n      <td>324</td>\n      <td>ENST00000257430</td>\n      <td>NM_000038.5</td>\n      <td>ENST00000257430</td>\n      <td>NM_000038.5</td>\n      <td>No</td>\n      <td>Yes</td>\n      <td>7</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>Yes</td>\n      <td>DP2.5, PPP1R46</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":2},{"cell_type":"code","source":"# Filtrating loaded genes by Vogelstein\nvogelstein_df = oncokb_df[oncokb_df['Vogelstein'] == 'Yes']\n# Saving names of the filtrated genes into a variable\nvogelstein_genes = vogelstein_df['Hugo Symbol'].tolist()\n# Changing for string to SQL\ngenes_str = \",\".join(f\"'{gene}'\" for gene in vogelstein_genes)\n\nprint(\"Number of genes after filtration: \", len(vogelstein_genes))\nprint(genes_str)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T17:15:47.150628Z","iopub.execute_input":"2025-06-28T17:15:47.150929Z","iopub.status.idle":"2025-06-28T17:15:47.161599Z","shell.execute_reply.started":"2025-06-28T17:15:47.150908Z","shell.execute_reply":"2025-06-28T17:15:47.160708Z"}},"outputs":[{"name":"stdout","text":"Number of genes after filtration:  125\n'ABL1','AKT1','ALK','AMER1','APC','AR','ARID1A','ASXL1','ATM','ATRX','AXIN1','BAP1','BCL2','BCOR','BRAF','BRCA1','BRCA2','CARD11','CBL','CDC73','CDH1','CDKN2A','CEBPA','CIC','CREBBP','CTNNB1','DAXX','DNMT3A','EGFR','EP300','ERBB2','EZH2','FBXW7','FGFR2','FGFR3','FLT3','FOXL2','GATA3','GNA11','GNAQ','GNAS','HNF1A','HRAS','IDH1','IDH2','JAK1','JAK2','JAK3','KDM5C','KDM6A','KIT','KMT2D','KRAS','MAP2K1','MAP3K1','MED12','MEN1','MET','MLH1','MPL','MSH2','MSH6','MYD88','NF1','NF2','NFE2L2','NOTCH1','NOTCH2','NPM1','NRAS','PAX5','PBRM1','PDGFRA','PIK3CA','PIK3R1','PPP2R1A','PRDM1','PTCH1','PTEN','PTPN11','RB1','RET','RNF43','SETD2','SF3B1','SMAD2','SMAD4','SMARCA4','SMARCB1','SMO','SOCS1','SPOP','STAG2','STK11','TET2','TNFAIP3','TP53','TSC1','U2AF1','VHL','WT1','ARID2','B2M','CASP8','CRLF2','CSF1R','FUBP1','GATA1','GATA2','H3-3A','H3C2','KMT2C','PHF6','RUNX1','SRSF2','ARID1B','CYLD','KLF4','NCOR1','SETBP1','SOX9','TRAF7','TSHR','ACVR1B','DNMT1'\n","output_type":"stream"}],"execution_count":3},{"cell_type":"markdown","source":"# **1.2** Mutation downloading","metadata":{}},{"cell_type":"code","source":"from google.cloud import bigquery\nfrom kaggle_secrets import UserSecretsClient\n\n# Setting BigQuery client\nclient = bigquery.Client()\n\n# SQL: Dowloading mutations only for filtrated genes\nquery = f\"\"\"\nSELECT *\nFROM `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_gdc_current`\nWHERE Hugo_Symbol IN ({genes_str})\nAND Variant_Type IN ('SNP', 'DEL', 'INS')\n\"\"\"\n\nquery_results = client.query(query)\nmutations_df = query_results.to_dataframe()\n\nprint(\"Found records of filtrated genes: \", len(mutations_df))\n","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"execution":{"iopub.status.busy":"2025-06-28T17:15:50.299979Z","iopub.execute_input":"2025-06-28T17:15:50.300613Z","iopub.status.idle":"2025-06-28T17:17:33.388007Z","shell.execute_reply.started":"2025-06-28T17:15:50.300588Z","shell.execute_reply":"2025-06-28T17:17:33.387035Z"}},"outputs":[{"name":"stdout","text":"Using Kaggle's public dataset BigQuery integration.\n","output_type":"stream"},{"name":"stderr","text":"/usr/local/lib/python3.11/dist-packages/google/cloud/bigquery/table.py:1727: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.\n  warnings.warn(\n","output_type":"stream"},{"name":"stdout","text":"Found records of filtrated genes:  48912\n","output_type":"stream"}],"execution_count":4},{"cell_type":"markdown","source":"# **1.3** Parquet format","metadata":{}},{"cell_type":"code","source":"# Saving filtrated records to Parquet\nmutations_df.to_parquet('filtered_mutations.parquet')","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T17:17:35.691942Z","iopub.execute_input":"2025-06-28T17:17:35.692218Z","iopub.status.idle":"2025-06-28T17:17:36.796803Z","shell.execute_reply.started":"2025-06-28T17:17:35.692198Z","shell.execute_reply":"2025-06-28T17:17:36.796075Z"}},"outputs":[],"execution_count":7},{"cell_type":"code","source":"# Downloading transcriptomic data from TCGA\nquery = f\"\"\"SELECT DISTINCT *\nFROM `isb-cgc-bq.TCGA.RNAseq_hg38_gdc_current`\nWHERE gene_name in ({genes_str});\n\"\"\"\nquery_job = client.query(query)\ndf_TCGA_rna = query_job.to_dataframe()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T17:17:36.798805Z","iopub.execute_input":"2025-06-28T17:17:36.799168Z","iopub.status.idle":"2025-06-28T17:25:32.505720Z","shell.execute_reply.started":"2025-06-28T17:17:36.799136Z","shell.execute_reply":"2025-06-28T17:25:32.504852Z"}},"outputs":[],"execution_count":8},{"cell_type":"code","source":"#and saving to parquet\ndf_TCGA_rna.to_parquet('expression.parquet')","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"# **1.4** Primary sites characteristics","metadata":{}},{"cell_type":"code","source":"print(df_TCGA_rna.columns)\n#how many recodrs in expression df\nprint(len(df_TCGA_rna))","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"code","source":"#checking how many records are for each primary site\nprint(df_TCGA_rna['primary_site'].value_counts())","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"# **Zad 2 features ingeneering**","metadata":{}},{"cell_type":"code","source":"#I had a problem dowloading this file here by kaggle. Chat said i could be a DNS problem\n# so I uploaded it to input manually\n!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T11:40:20.805812Z","iopub.execute_input":"2025-06-28T11:40:20.806164Z","iopub.status.idle":"2025-06-28T11:40:52.992691Z","shell.execute_reply.started":"2025-06-28T11:40:20.806133Z","shell.execute_reply":"2025-06-28T11:40:52.991602Z"}},"outputs":[{"name":"stdout","text":"--2025-06-28 11:40:20--  http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\nResolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... failed: Temporary failure in name resolution.\nwget: unable to resolve host address ‘hgdownload.soe.ucsc.edu’\n","output_type":"stream"}],"execution_count":19},{"cell_type":"code","source":"!gunzip \"/kaggle/input/hg38_dataset/hg38.fa.gz\"","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T14:36:02.513407Z","iopub.execute_input":"2025-06-28T14:36:02.514170Z","iopub.status.idle":"2025-06-28T14:36:02.636930Z","shell.execute_reply.started":"2025-06-28T14:36:02.514134Z","shell.execute_reply":"2025-06-28T14:36:02.635955Z"}},"outputs":[{"name":"stdout","text":"gzip: /kaggle/input/hg38_dataset/hg38.fa.gz: No such file or directory\n","output_type":"stream"}],"execution_count":5},{"cell_type":"code","source":"import Bio\nprint(Bio.__version__)\n","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"code","source":"!pip install biopython","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T14:12:53.030269Z","iopub.execute_input":"2025-06-28T14:12:53.030556Z","iopub.status.idle":"2025-06-28T14:16:15.518129Z","shell.execute_reply.started":"2025-06-28T14:12:53.030532Z","shell.execute_reply":"2025-06-28T14:16:15.517258Z"}},"outputs":[{"name":"stdout","text":"\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7be1a2606890>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7be1a25cadd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7be1a2685110>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7be1a25e9a90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7be1a25f4390>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/biopython/\u001b[0m\u001b[33m\n\u001b[0m\u001b[31mERROR: Could not find a version that satisfies the requirement biopython (from versions: none)\u001b[0m\u001b[31m\n\u001b[0m\u001b[31mERROR: No matching distribution found for biopython\u001b[0m\u001b[31m\n\u001b[0m","output_type":"stream"}],"execution_count":1},{"cell_type":"code","source":"import Bio\nfrom Bio import SeqIO\n\nfasta_sequences = SeqIO.to_dict(SeqIO.parse(\"/kaggle/input/hg38_dataset/hg38.fa\", \"fasta\"))\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T14:36:35.560334Z","iopub.execute_input":"2025-06-28T14:36:35.560655Z","iopub.status.idle":"2025-06-28T14:36:35.572110Z","shell.execute_reply.started":"2025-06-28T14:36:35.560628Z","shell.execute_reply":"2025-06-28T14:36:35.570928Z"}},"outputs":[{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m/tmp/ipykernel_35/1491488935.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mfasta_sequences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/kaggle/input/hg38_dataset/hg38.fa\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'Bio'"],"ename":"ModuleNotFoundError","evalue":"No module named 'Bio'","output_type":"error"}],"execution_count":7},{"cell_type":"code","source":"def get_motif(row):\n    ref = row['Reference_Allele']\n    alt = row['Tumor_Seq_Allele2']\n    context = row['context']  # np. ACA, TTC itp.\n    return f\"{context[0]}[{ref}>{alt}]{context[2]}\"\n\nmutations_df['motif'] = mutations_df.apply(get_motif, axis=1)\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-28T11:17:20.590045Z","iopub.execute_input":"2025-06-28T11:17:20.590342Z","iopub.status.idle":"2025-06-28T11:17:20.940699Z","shell.execute_reply.started":"2025-06-28T11:17:20.590323Z","shell.execute_reply":"2025-06-28T11:17:20.939235Z"}},"outputs":[{"name":"stdout","text":"The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.\n","output_type":"stream"},{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3805\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3806\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32mindex.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n","\u001b[0;32mindex.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n","\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n","\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n","\u001b[0;31mKeyError\u001b[0m: 'context'","\nThe above exception was the direct cause of the following exception:\n","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","\u001b[0;32m/tmp/ipykernel_35/865009820.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;34mf\"{context[0]}[{ref}>{alt}]{context[2]}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mmutations_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'motif'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmutations_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_motif\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m  10372\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10373\u001b[0m         )\n\u001b[0;32m> 10374\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"apply\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m  10375\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10376\u001b[0m     def map(\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    914\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    915\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 916\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    917\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    918\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1061\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1062\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1063\u001b[0;31m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1064\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1065\u001b[0m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_numba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1079\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1080\u001b[0m                 \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1081\u001b[0;31m                 \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1082\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m                     \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/tmp/ipykernel_35/865009820.py\u001b[0m in \u001b[0;36mget_motif\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Reference_Allele'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0malt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Tumor_Seq_Allele2'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mcontext\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'context'\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# np. ACA, TTC itp.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;34mf\"{context[0]}[{ref}>{alt}]{context[2]}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1120\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mkey_is_scalar\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1121\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1122\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1123\u001b[0m         \u001b[0;31m# Convert generator to list before going through hashable part\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1235\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1236\u001b[0m         \u001b[0;31m# Similar to Index.get_value, but we do not fall back to positional\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1237\u001b[0;31m         \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1238\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1239\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3810\u001b[0m             ):\n\u001b[1;32m   3811\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mInvalidIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3812\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3813\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3814\u001b[0m             \u001b[0;31m# If we have a listlike key, _check_indexing_error will raise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mKeyError\u001b[0m: 'context'"],"ename":"KeyError","evalue":"'context'","output_type":"error"}],"execution_count":10}]}